File: cpp/source/state/device/Connected.cpp
#include "state/device/Connected.hpp"

#include "state/device/connected/Ready.hpp"

#include "backend/ClientBackend.hpp"

using namespace fm;
using namespace fm::state;
using namespace fm::state::device;

using event::input::user::UserEvent;
using event::input::connection::ConnectionEvent;

using event::output::FacadeEvent;

Connected::Connected(IState& state) :
    IState(state),
    internalState(std::make_unique<connected::Ready>(*this))
{
}

std::unique_ptr<IState> Connected::start()
{
    backend.getHeartbeatHandler().start();
    internalState->start();
    listener.onEvent(std::make_shared<FacadeEvent>(FacadeEvent::ATTACHED));
    return nullptr;
}

std::unique_ptr<IState> Connected::handleUserEvent(const UserEvent& event)
{
    return onNewState(internalState->handleUserEvent(event));
}

std::unique_ptr<IState> Connected::handleConnectionEvent(const ConnectionEvent& event)
{
    return onNewState(internalState->handleConnectionEvent(event));
}

std::string Connected::toString() const
{
    return "Connected::" + internalState->toString();
}

std::unique_ptr<IState> Connected::onNewState(std::unique_ptr<IState> newState)
{
    while (nullptr != newState.get())
    {
        listener.trace("Connected transition: " + internalState->toString() + " -> " + newState->toString());
        internalState.swap(newState);
        newState.reset(internalState->start().release());
    }
    return internalState->createOuterState();
}

File: java/src/main/java/com/fleetmgr/sdk/client/state/device/Connected.java
package com.fleetmgr.sdk.client.state.device;

import com.fleetmgr.sdk.client.event.input.Event;
import com.fleetmgr.sdk.client.event.input.connection.ConnectionEvent;
import com.fleetmgr.sdk.client.event.input.user.UserEvent;
import com.fleetmgr.sdk.client.event.output.facade.FacadeEvent;
import com.fleetmgr.sdk.client.state.device.connected.Ready;
import com.fleetmgr.sdk.client.state.device.connected.Recovering;
import com.fleetmgr.sdk.client.state.device.connected.Released;
import com.fleetmgr.sdk.client.traffic.Channel;
import com.fleetmgr.sdk.system.machine.BaseState;

import java.util.Optional;
import java.util.logging.Logger;

/**
 * Created by: Bartosz Nawrot
 * Date: 23.09.2018
 * Description:
 */
public class Connected extends BaseState {

    private static final Logger logger = Logger.getLogger(Connected.class.getName());


    private BaseState internalState;

    Connected(BaseState state) {
        super(state);
        this.internalState = new Ready(this);
    }

    @Override
    public void start() {
        internalState.start();
        backend.getHeartbeatHandler().start();
        listener.onEvent(new FacadeEvent(FacadeEvent.Type.ATTACHED));
    }

    @Override
    public Optional<BaseState> onUserEvent(UserEvent event) {
        return handle(event);
    }

    @Override
    public Optional<BaseState> onConnectionEvent(ConnectionEvent event) {
        return handle(event);
    }

    private Optional<BaseState> handle(Event event) {
        Optional<BaseState> newState = internalState.handleEvent(event);
        boolean wasRecovering = internalState instanceof Recovering;

        while (newState.isPresent()) {
            logger.info("Connected transition: " + toString() + " -> Connected." + newState.toString());
            internalState = newState.get();

            internalState.start();
            newState = internalState.getInitialState();
        }

        if (internalState instanceof Released) {
            return Optional.of(new Disconnecting(this, wasRecovering));
        }

        return Optional.empty();
    }

    @Override
    public String toString() {
        return "Connected." + internalState;
    }
}
